src/java.base/share/classes/java/lang/Module.java

```diff
@@ -258,24 +258,41 @@ public ModuleLayer getLayer() {
     /**
      * Update this module to allow access to restricted methods.
      */
-    synchronized Module implAddEnableNativeAccess() {
-        enableNativeAccess = true;
+    Module implAddEnableNativeAccess() {
+        AccessHolder.trySetEnableNativeAccess(this);
         return this;
     }
 
     /**
      * Returns {@code true} if this module can access
      * <a href="foreign/package-summary.html#restricted"><em>restricted</em></a> methods.
      *
-     * @since 20
-     *
      * @return {@code true} if this module can access <em>restricted</em> methods.
+     * @since 20
      */
-    @PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+    @PreviewFeature(feature = PreviewFeature.Feature.FOREIGN)
     public boolean isNativeAccessEnabled() {
         Module target = moduleForNativeAccess();
-        synchronized(target) {
-            return target.enableNativeAccess;
+        return AccessHolder.isNativeAccessEnabled(target);
+    }
+
+    // This class is used to be able to bootstrap without using Unsafe
+    // in the outer Module class as that would create a circular initializer dependency
+    private static final class AccessHolder {
+
+        private AccessHolder() {}
+
+        private static final Unsafe UNSAFE = Unsafe.getUnsafe();
+        private static final long FIELD_OFFSET = UNSAFE.objectFieldOffset(Module.class, "enableNativeAccess");
```

Comment:
Using Unsafe and CAS'ing the enableNativeAccess field looks okay.  The name "AccessHolder" looks too general here as there is a lot of access going on in this class, maybe NativeAccessHolder will work.

It would be good if you could try to keep changes to this code consistent with the existing style if possible. In this case, the other inner class uses /** .. */ style comment for the class description. I think would I put the constants at the top of the this class.